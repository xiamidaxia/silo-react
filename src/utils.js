const objKeys = Object.keys

export const reduce = (obj, fn, origin) => objKeys(obj).reduce((res, key) => fn(res, obj[key], key), origin)

export const mapValues = (obj, fn) => reduce(obj, (res, val, key) => Object.assign(res, { [key]: fn(val, key) }), {})
