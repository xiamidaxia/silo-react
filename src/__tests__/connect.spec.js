import React from 'react'
import renderer from 'react-test-renderer'
import { createSiloStore, connect, Provider } from '../index'
import { addTestPath } from './utils'

describe('connect', () => {
  it('create connect', () => {
    const Demo = connect('test')(({ state, getters, setters }) => (
      <div>
        <span>count: {state.count}: {getters.getCount()}</span>
        <button onClick={() => setters.change()}>button</button>
      </div>
    ))
    const store = createSiloStore()
    addTestPath(store)
    const component = renderer.create(
      <Provider store={store}>
        <Demo />
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
    tree.children[1].props.onClick()
    expect(component.toJSON()).toMatchSnapshot()
  })
  it('connect update', () => {
    let renderTimes = 0
    const Demo = connect('test')(() => {
      renderTimes++
      return <div>demo</div>
    })
    const store = createSiloStore()
    addTestPath(store)
    const component = renderer.create(
      <Provider store={store}>
        <Demo />
      </Provider>
    )
    expect(renderTimes).toBe(1)
    store.exec('set:test/noChange')
    expect(renderTimes).toBe(1)
    store.exec('set:test/change')
    store.exec('set:test/change')
    expect(renderTimes).toBe(3)
    store.exec('action:test/mulitiChange')
    expect(renderTimes).toBe(4)
    component.unmount()
    store.exec('set:test/change')
    expect(renderTimes).toBe(4)
  })
})
