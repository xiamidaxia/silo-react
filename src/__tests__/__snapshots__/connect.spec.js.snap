// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect create connect 1`] = `
<div>
  <span>
    count: 
    0
    : 
    0
  </span>
  <button
    onClick={[Function]}
  >
    button
  </button>
</div>
`;

exports[`connect create connect 2`] = `
<div>
  <span>
    count: 
    1
    : 
    1
  </span>
  <button
    onClick={[Function]}
  >
    button
  </button>
</div>
`;
